pyro/poutine/runtime.py

```diff
@@ -64,26 +63,26 @@ class InferDict(TypedDict, total=False):
     _markov_depth: int
 
 
-class Message(TypedDict, total=False):
+class Message(TypedDict, Generic[P, T], total=False):
     type: str
     name: Optional[str]
-    fn: Union[Callable, TorchDistributionMixin]
+    fn: Callable[P, T]
     is_observed: bool
     args: Tuple
     kwargs: Dict
-    value: Optional[torch.Tensor]
+    value: Optional[T]
```

Comment:
I've learned this neat trick with `Generic` and `TypeVar` where the type of `value` can be inferred from the `Callable` signature. I also fixed `effectful` so that it gives the correct signature for the decorated function when the return type is diferent from `torch.Tensor` (e.g. `reparam_messenger._get_init_messengers`).